TLM/TLM/Util/RoundaboutMassEdit.cs

```diff
@@ -333,6 +333,7 @@ public class RoundaboutMassEdit {
             bool preferLeft) {
             ArrowDirection preferDir = preferLeft ? ArrowDirection.Left : ArrowDirection.Right;
 
+            HashSet<ushort> hashSet = new HashSet<ushort>();
             List<ushort> sortedSegList = new List<ushort>();
 
```

Comment:
does GetNodeSegmentIds return the same segment ID twice?

Reply:
I don't remember, but I doubt. 
When I was debugging it, segment ID has been inserted to the front of sortedlist each iteration(condition was always true), basically list looked like this:
```cs
[segmentID_2, segmentID_2, segmentID_2 ,segmentID_2, segmentID_2, (etc...), segmentID_1]
```
_Iterating list and inserting items to it at the same time is not the best idea._

Simplest possible setup, seem to be related with directions of one-way connection, but I didn't have time to dig deep in your algorithm, hence this simplest possible "fix"

![image](https://user-images.githubusercontent.com/19638970/123012484-9d3bf300-d3c2-11eb-99b0-2c654d34501f.png)
_(credits to Chamëleon)_

Reply:
the only way this can fail in a way ur hashset fixes it is if netService.GetNodeSegmentIds returns an infinite number of segments.
I think we need to dig a bit deeper to find out about more potential problems too

Reply:
this could be a regression by @DaEgi01  enumerable

Reply:
possible

Reply:
I think this is what I mean to do:
```cs
            foreach (var nodeSegmentId in netService.GetNodeSegmentIds(headNodeId, ClockDirection.CounterClockwise)) {
                if (!IsPartofRoundabout(nodeSegmentId, segmentId, headNodeId)) {
                    continue;
                }
                if (segEndMan.GetDirection(segmentId, nodeSegmentId, headNodeId) == dir) {
                    for (int i = 0; i < sortedSegList.Count; ++i) {
                        if (segEndMan.GetDirection(nodeSegmentId, sortedSegList[i], headNodeId) == preferDir) {

                            sortedSegList.Insert(i, nodeSegmentId);
                            goto nextLoop;
                        }
                    }

                    sortedSegList.Add(nodeSegmentId);
                }
                nextLoop:;
            }
```

Reply:
in any case `segEndMan.GetDirection(nodeSegmentId, sortedSegList[i], headNodeId) == preferDir` os also wrong.
I wanted to detect small angles. but that does not happen in here for the most part.
it would be better for me to sort by angle afterwards. makes the code much simpler.